python/vespa/vespa/package.py

```diff
@@ -468,3 +490,160 @@ def deploy(self, disk_folder: str, container_memory: str = "4G"):
             port=self.local_port,
             deployment_message=deployment_message,
         )
+
+
+class VespaCloud(object):
+    def __init__(self, tenant: str, application: str, key_location: str) -> None:
+        """
+        Deploy application to the Vespa Cloud (cloud.vespa.ai)
+
+        :param tenant: Tenant name registered in the Vespa Cloud.
+        :param application: Application name registered in the Vespa Cloud.
+        :param key_location: Location of the private key used for signing HTTP requests to the Vespa Cloud.
+        """
+        self.tenant = tenant
+        self.application = application
+        self.api_key = self.read_private_key(key_location)
+        self.api_public_key_bytes = standard_b64encode(self.api_key.public_key().public_bytes(serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo))
+        self.data_key, self.data_certificate = self.create_certificate_pair()
+        self.data_cert_file = self.write_private_key_and_cert(self.data_key, self.data_certificate)
+        self.connection = http.client.HTTPSConnection('api.vespa-external.aws.oath.cloud', 4443)
+
+    @staticmethod
+    def read_private_key(key_location: str) -> ec.EllipticCurvePrivateKey:
+        with open(key_location, 'rb') as key_data:
+            key = serialization.load_pem_private_key(key_data.read(), None, default_backend())
+            if not isinstance(key, ec.EllipticCurvePrivateKey):
+                raise TypeError("Key at " + key_location + " must be an elliptic curve private key")
+            return key
+
+    @staticmethod
+    def write_private_key_and_cert(key: ec.EllipticCurvePrivateKey, cert: x509.Certificate) -> (str, str):
+        cert_file = tempfile.NamedTemporaryFile('wt')
```

Comment:
This file can only be read by the user writing it, and is deleted immediately when closed. The `VespaCloud` closes this file on _its_ close, and is meant to be used like
```python
with VespaCloud('tenant', 'application', '/path/to/api_key') as cloud:
    ...
```

Reply:
Is there a way to use an in-memory file object instead of a temporary on-disk file?

Reply:
Absolutely not, unfortunately. It's been filed as a "bug" for 8 years, but nothing has happened.

Reply:
There were even made PRs to address the issue, but the maintainers rejected them because they wanted a "perfect" solution, but this solution has, of course, not arrived, nor does it seem like it will.

Reply:
How pythonic

Reply:
Another option is to use 
```
vespa_cloud = VespaCloud("vespa-team", "ms-marco", "/Users/tmartins/sample_application/tmartins.vespa-team.pem")
app = vespa_cloud.deploy('from-notebook', app_package)
```
Then continue  doing all the operation like `app.feed_data_point(...)` and then use `app.close()` when you are done.

Does that makes sense @jonmv?

Reply:
Do you mean if it makes sense to move generation of data place key/cert to when you deploy, so it lives only while  a particular deployment (`Vespa` object) is in scope? I think that makes sense, but there's not much practical difference. You can avoid keeping the `VespaCloud` object in scope, but that's it. 
And for the certificates, it's possible to just never `.close()` the object, of course. But you'd have a hard time finding the certificate file without printing its name somewhere ... :)

Reply:
```python
with VespaCloud(...) as cloud:
    with cloud.deploy(...) as app:
        # both variables are in scope now, as you can't have the Vespa in scope without the VespaCloud anyway
        # and the VespaCloud is intended to be closed regardless of where the certs are, due to its HTTPConnection
```

Reply:
 I guess you could perhaps do 
```python
with VespaCloud(...) as cloud:
    app = cloud.deploy(...)

with app:
    # do stuff
```

Reply:
Anyway, these are all details. Do whatever you think is better :)